diff --git a/node_modules/@videosdk.live/expo-ios-screen-share/build/withIosBroadcastExtension.js b/node_modules/@videosdk.live/expo-ios-screen-share/build/withIosBroadcastExtension.js
index 3ebc704..a638c69 100644
--- a/node_modules/@videosdk.live/expo-ios-screen-share/build/withIosBroadcastExtension.js
+++ b/node_modules/@videosdk.live/expo-ios-screen-share/build/withIosBroadcastExtension.js
@@ -311,7 +311,7 @@ const addXCConfigurationList = (proj, { extensionBundleIdentifier, currentProjec
         INFOPLIST_FILE: `${extensionName}/Info.plist`,
         INFOPLIST_KEY_CFBundleDisplayName: `${extensionName}`,
         INFOPLIST_KEY_NSHumanReadableCopyright: (0, utils_1.quoted)(''),
-        IPHONEOS_DEPLOYMENT_TARGET: '14.0',
+        IPHONEOS_DEPLOYMENT_TARGET: '15.1',
         LD_RUNPATH_SEARCH_PATHS: (0, utils_1.quoted)('$(inherited) @executable_path/Frameworks @executable_path/../../Frameworks'),
         MARKETING_VERSION: marketingVersion,
         MTL_FAST_MATH: 'YES',
@@ -348,7 +348,7 @@ const addXCConfigurationList = (proj, { extensionBundleIdentifier, currentProjec
     const xCConfigurationList = proj.addXCConfigurationList(buildConfigurationsList, 'Release', `Build configuration list for PBXNativeTarget ${(0, utils_1.quoted)(extensionName)}`);
     // update other build properties
     proj.updateBuildProperty('ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES', 'YES', null, proj.getFirstTarget().firstTarget.name);
-    proj.updateBuildProperty('IPHONEOS_DEPLOYMENT_TARGET', '14.0');
+    proj.updateBuildProperty('IPHONEOS_DEPLOYMENT_TARGET', '15.1');
     return xCConfigurationList;
 };
 const addProductFile = (proj, extensionName, groupName) => {
